### DISCUSSION TITLE
Converting tables extracted from MDB to GeoDataFrames

### DISCUSSION THREAD

**Discussion by awa5114** [geopandas/geopandas#3465] [Q&A]
I have an MDB Access database which I believe originates from an Esri GeoDatabase. I do not have access to ArcGis and would like to read some tables from this database using geopandas. Here's what I came up with so far: 

```
db = "../db.mdb"
data = {}

conn_str = f'DRIVER={{Microsoft Access Driver (*.mdb, *.accdb)}}; DBQ={db};'
connection_url = sa.engine.URL.create(
    "access+pyodbc",
    query={"odbc_connect": conn_str}
)
engine = sa.create_engine(connection_url)

for table_name in sa.inspect(engine).get_table_names():
    query = f"select * from {table_name}"
    df = pd.read_sql(query, engine)
    cols_upper = [c.upper() for c in df.columns]
    if "SHAPE" in cols_upper:
        df.columns = cols_upper
        data[table_name] = gpd.GeoDataFrame(df, geometry="SHAPE")
```

Unfortunately, geopandas doesn't like the format for the geometry column and I get the following error: 

> TypeError: Input must be valid geometry objects: b'\x01\x00\x00\x00\x00\xc0\xaag [\x0eA\x00[p\xd3\xe2H\xfe@' 

How can I convert the SHAPE columns to a format acceptable by geopandas?

**Comment by awa5114** [comment]
It turns out the MDB was an ESRI personal geodatabase and this is more or less supported by OGR, see [doc](https://gdal.org/en/latest/drivers/vector/pgeo.html)

I was able to convert this to a GPKG using OGR2OGR:

```
ogr2ogr -f GPKG db.gpkg --config PGEO_DRIVER_TEMPLATE="DRIVER={Microsoft Access Driver (*.mdb, *.accdb)};DBQ=C:\Users\admin\Desktop\TEMP\projects\314\db.mdb;" db.mdb
```

I think I should be able to go from GPKG to geodataframe relatively easily now

**Comment by m-richards** [comment]
Glad you found a way via ogr2ogr but this should be possible directly with geopandas

Did you try using `geopandas.read_file` on the mdb database directly? Otherwise you may also be able to use use `geopandas.read_postgis` - as although the name does not suggest it, this can be used to read from other file databases like sqlite.

for the snippet you've given, you'd need to convert the WKB geometry explicitly:
```python
...
for table_name in sa.inspect(engine).get_table_names():
    query = f"select * from {table_name}"
    df = pd.read_sql(query, engine)
    cols_upper = [c.upper() for c in df.columns]
    if "SHAPE" in cols_upper:
        df.columns = cols_upper
        data[table_name] = gpd.GeoDataFrame(df.drop(columns=["SHAPE"]), geometry=gpd.GeoSeries.from_wkb(df["SHAPE"]))
```
though you'd likely need a secondary query to retrieve the SRID/CRS information from however that is stored in and mdb.

**Reply by awa5114** [reply]
I'm pretty sure anything coming from ESRI (Shapefiles, geodatabases) is not in WKB format. So `from_wkb` will not work...

**Reply by m-richards** [reply]
sorry, missed the ESRI part on re-reading this. Though it would still be good to know if using `read_file` directly works with your data
